tools/e2fsck-lcd/e2fsck-lcd.8
.\" $Id$
.\"
.\" Copyright (C) 2004  Martin Michlmayr <tbm@cyrius.com>
.\" This manual is freely distributable under the terms of the GPL.
.\" It was originally written for Debian GNU/Linux (but may be used
.\" by others).
.\"

.TH E2FSCK-LCD 8 "July 2004"

.SH NAME
e2fsck-lcd \- output file system check progress information on the Cobalt LCD

.SH SYNOPSIS
.B e2fsck-lcd

.SH DESCRIPTION
.PP
.B e2fsck-lcd
is a program which outputs file system check progress information on the
Cobalt LCD.  It will call the real
.B fsck.ext2
and
.B fsck.ext3
programs to perform the file checking functions and output progress
information on the Cobalt LCD.

If called as
.B e2fsck-lcd
the program will pass command line arguments to
.B fsck.ext2
to perform the file system check. If called by any other name it will
use the called name with
.B \.distrib
appended as the name of the file system check program.

.SH AUTHOR
.PP
Peter Horton <pdh@colonel-panic.org>



CHANGES
# $Id$

CoLo 1.12 (xxxx-xx-xx)
----------------------

*	Added 'menu' command for implementing LCD menus in scripts. Makes it
	possible to choose from a list of multiple kernels/configurations.

*	Fixed a couple of off by one errors in the 'boot' command.

*	Added 'console disable' to the non-volatile flags. This defers initialising
	the serial port until the boot shell is entered and should make it possible
	to use CoLo on units which don't have a serial port. This flag will be set
	by default on new installs, but will be clear on units which have a recent
	CoLo install already. Might also be useful where the serial port is used
	for a modem etc.

*	Added man pages for 'paneld' and 'e2fsck-lcd'.
	[Martin Michlmayr]

CoLo 1.11 (2004-07-28)
----------------------

*	Build all the tools.

*	You can now break out of the boot menu via the console.

*	Added man pages for 'colo-perm' and 'md5rom'.
	[Martin Michlmayr]

*	Added 'e2fsck-lcd', file check progress bar on LCD.

*	Fixed LCD second row offset (all over).

*	Make shell refuse to load chain loader.

*	Added simple front panel daemon 'paneld'.

*	Added 'restrict' command to shell.

CoLo 1.10 (2004-05-23)
----------------------

*	Fixed loading of files with sparse blocks.

*	Added signature to Flash, and 'colo-perm' to test for it.

*	Changed PCI latency registers to 64 (matches kernel default).

*	Added NFS support ('nfs' command).

*	Removed debugging output from chain loader. Chain loader now uses LCD for
 	error messages.

*	Added command 'serial' for changing serial baud rate. Baud rate setting is
	non-volatile.

CoLo 1.9 (2004-05-02)
---------------------

*	CTRL-C now acts as a break key (in addition to SPACE).

*	Comments are now only supported in scripts.

*	The 'script' command has been changed to execute a script already loaded
 	into memory via the 'load', 'tftp' or 'download' commands.

	** NOTE **

		Scripts must now begin with the line '#:CoLo:#'.

*	Automatic insertion of the kernel 'initrd=' command line argument has been
 	removed. All initrd parameters are now available in variables.

*	Made use of non-volatile storage on RTC.

*	Renamed 'dflags' command to 'nvflags' and made the flags non-volatile.

*	Added boot menu and 'boot' command.

	*note*

		If you do not want the boot menu to appear every boot use the command
		'boot default 1' (see README.shell).

*	Added 'variable' command.

*	Implemented variable substition in shell command lines.

*	DHCP now sets variables with boot information.

*	Added 'lcd' command.

CoLo 1.8 (2004-04-26)
---------------------

*	Added 'pci' command.

*	Added 'ping' command.

*	Added 'tftp' command.

*	Added network stack (the 'net' command).

*	Added man pages for 'flash-tool' and 'putlcd'.
	[Martin Michlmayr]

CoLo 1.6 (2004-04-18)
---------------------

*	Added support for RAID partitions.
	[Christian Mock]

*	The layout of the source tree has been tidied up.

*	Stage 2 is now loaded at the top of RAM rather than the bottom. We can now
	load kernels linked to run at 0x80000000. This is a big change, but it all
	seems to work.

	Unfortunately the chain loader can no longer detect if it's being loaded
	from "CoLo".

*	Build output files are now 'colo-rom-image.bin' and 'colo-chain.elf'.

*	Boot loader now sources the script '/boot/default.colo' by default, rather
	than '/boot/default.boot'.

*	Fixed stupid error in linking of chain loader.

CoLo 1.5 (2004-04-11)
---------------------

*	The boot loader now has a name, "CoLo".
	[suggested by Kumba]

*	Added support for loading initial RamDisk. A second path passed to the
	'load' command is interpreted as a file to be loaded as an initial RamDisk.
	The 2004-04-11 2.4.x kernel patch includes the kernel side of this
	functionality.
	[suggested by Martin Michlmayr]

*	The chain loader now halts with an error message if it's loaded from "CoLo"
	rather than from the original Cobalt loader.

# vi:set ts=3 sw=3 tw=78 ai:


tools/e2fsck-lcd/e2fsck-lcd.c
/*
 * (C) P.Horton 2004
 *
 * $Id$
 *
 * This code is covered by the GNU General Public License. For details see the file "COPYING".
 */

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <assert.h>
#include <sys/wait.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/param.h>

#define APP_NAME						"e2fsck-lcd"
#define EXTENSION						".distrib"
#define CHILD_DEFAULT				"fsck.ext2"
#define ERROR_EXIT					8						/* e2fsck - operational error */

#define LCD_WIDTH						16

#define LCD_PHYS_ADDR				0x1f000000

#define LCD_READ(r)					(lcd[!!(r)*4]>>24)
#define LCD_WRITE(r,v)				do{lcd[!!(r)*4]=(unsigned)(v)<<24;}while(0)

#define LCD_BUSY						(1 << 7)
#define LCD_CLEAR						0x01
#define LCD_DDRAM_ADDR				0x80
#define LCD_ROW_OFFSET				0x40

static const char spaces[] = { [0 ... (LCD_WIDTH - 1)] = ' ' };
static volatile uint32_t *lcd;
static unsigned thumb;

static void lcd_write(int reg, unsigned val)
{
	while(LCD_READ(0) & LCD_BUSY)
		usleep(1);
	
	usleep(10);

	LCD_WRITE(reg, val);
}

/*
 * clear display
 */
static void disp_clear(void)
{
	lcd_write(0, LCD_CLEAR);
}

/*
 * write text to display
 */
static void disp_text(const char *str, unsigned size)
{
	unsigned indx;

	for(indx = 0; str[indx] && indx < size; ++indx)
		lcd_write(1, str[indx]);
}

/*
 * move display cursor
 */
static void disp_cursor_move(unsigned row, unsigned col)
{
	lcd_write(0, LCD_DDRAM_ADDR | (row * LCD_ROW_OFFSET + col));
}

/*
 * initialise output
 */
static void disp_init(const char *devn)
{
	int fd;

	fd = open("/dev/mem", O_RDWR | O_SYNC);
	if(fd == -1) {
		fprintf(stderr, APP_NAME ": failed to open \"/dev/mem\" (%s)\n", strerror(errno));
		return;
	}

	lcd = mmap(0, 32, PROT_READ | PROT_WRITE, MAP_SHARED, fd, LCD_PHYS_ADDR);
	if(lcd == MAP_FAILED) {
		fprintf(stderr, APP_NAME ": failed to map \"/dev/mem\" (%s)\n", strerror(errno));
		lcd = NULL;
		close(fd);
		return;
	}

	close(fd);

	disp_clear();

	if(devn) {

		disp_cursor_move(0, 0);
		disp_text(devn, LCD_WIDTH - 7);
	}

	disp_cursor_move(0, LCD_WIDTH - 4);
	disp_text("0.0%", -1);

	disp_cursor_move(1, 0);
	disp_text("[>", -1);
	disp_cursor_move(1, LCD_WIDTH - 1);
	disp_text("]", -1);

	thumb = 0;
}

/*
 * finish output
 */
static void disp_cleanup(void)
{
	usleep(500 * 1000);

	if(lcd)
		disp_clear();

	fputs("       \r", stdout);
	fflush(stdout);
}

/*
 * update output with new progress information
 */
static void disp_update(int pass, unsigned where, unsigned limit)
{
	static char bar[] = { [0 ... (LCD_WIDTH - 1)] = '=' };
	static char text[16];
	unsigned prog;

	if(pass < 1 || pass > 5)
		return;

	prog = (pass - 1) * 200 + (where * 200 + limit / 2) / limit;

	snprintf(text, sizeof(text), "%3u.%u%%", prog / 10, prog % 10);
	text[sizeof(text) - 1] = '\0';

	if(lcd) {

		disp_cursor_move(0, LCD_WIDTH - 6);
		disp_text(text, 6);

		prog = ((LCD_WIDTH - 2) * prog + 500) / 1000;
		if(prog > thumb) {
			disp_cursor_move(1, thumb + 1);
			disp_text(bar, prog - thumb);
			thumb = prog;
			if(thumb < LCD_WIDTH - 2)
				disp_text(">", -1);
		}
	}

	printf(" %s\r", text);
	fflush(stdout);
}

/*
 * read progress information from pipe
 */
static void monitor(int fd, const char *devn)
{
	static char ibuf[8192];

	unsigned fill, pass, where, limit;
	int cntl, nread, skip, init;
	char pad, *ptr;
	fd_set set;

	cntl = fcntl(fd, F_GETFL);
	if(cntl == -1 || fcntl(fd, F_SETFL, cntl | O_NONBLOCK)) {
		fprintf(stderr, APP_NAME ": failed to set pipe to non-blocking (%s)\n", strerror(errno));
		return;
	}

	if(devn) {
		ptr = strrchr(devn, '/');
		if(ptr)
			devn = ptr + 1;
	}

	FD_ZERO(&set);

	for(init = 0, fill = 0, skip = 0;;) {

		FD_SET(fd, &set);

		if(select(fd + 1, &set, NULL, NULL, NULL) == -1 && errno != EINTR) {
			fprintf(stderr, APP_NAME ": failed waiting on pipe (%s)\n", strerror(errno));
			return;
		}

		nread = read(fd, ibuf + fill, sizeof(ibuf) - fill);
		if(!nread)
			break;

		if(nread == -1) {
			if(errno != EINTR && errno != EAGAIN) {
				fprintf(stderr, APP_NAME ": failed reading from pipe (%s)\n", strerror(errno));
				return;
			}
			continue;
		}

		fill += nread;

		for(;;) {

			ptr = strchr(ibuf, '\n');
			if(!ptr) {

				if(fill == sizeof(ibuf)) {

					if(!skip)
						fputs(APP_NAME ": line too long\n", stderr);

					skip = 1;
					fill = 0;
				}

				break;
			}

			*ptr++ = '\0';

			if(!skip) {

				if(sscanf(ibuf, " %u %u %u %c", &pass, &where, &limit, &pad) == 3) {

					if(!init) {
						disp_init(devn);
						init = 1;
					}

					disp_update(pass, where, limit);

				} else

					fputs(APP_NAME ": line invalid format\n", stderr);
			}

			skip = 0;

			fill -= ptr - ibuf;
			memmove(ibuf, ptr, fill);
		}
	}

	if(init > 0)
		disp_cleanup();
}

int main(int argc, char *argv[])
{
	static char path[PATH_MAX] = CHILD_DEFAULT;

	char *ptr, *end, *desc, *devn, *argp[argc + 1];
	int next, retn, tube[2];
	pid_t child, dead;
	struct stat info;
	unsigned indx;

	/* work out executable to spawn */

	ptr = strrchr(argv[0], '/') + 1;
	if(ptr == (char *) NULL + 1)
		ptr = argv[0];

	if(strcmp(ptr, APP_NAME)) {
		ptr = path + strlen(argv[0]);
		if(ptr + sizeof(EXTENSION) > path + sizeof(path)) {
			fputs(APP_NAME ": executable path too long\n", stderr);
			return ERROR_EXIT;
		}
		strcpy(path, argv[0]);
		strcpy(ptr, EXTENSION);
	}

	argp[0] = path;
	for(indx = 1; indx < argc; ++indx)
		argp[indx] = argv[indx];
	argp[indx] = NULL;

	/* find argument of -C option and device node */

	desc = NULL;
	devn = NULL;
	next = 0;

	for(indx = 0;; ++indx) {

		ptr = argp[indx];
		if(!ptr)
			break;

		if(next) {
			next = 0;
		} else if(*ptr == '-') {
			ptr = strchr(ptr, 'C');
			if(!ptr)
				continue;
			next = !*++ptr;
			if(next)
				continue;
		} else if(stat(ptr, &info) == 0 && S_ISBLK(info.st_mode))
			devn = ptr;

		if(!desc) {
			strtoul(ptr, &end, 0);
			if(end > ptr && !*end)
				desc = ptr;
		}
	}

	if(desc) {

		/* start e2fsck as child */

		close(3);
		close(4);

		if(pipe(tube))

			fprintf(stderr, APP_NAME ": failed to create pipe (%s)\n", strerror(errno));

		else if(tube[1] < 10) {

			child = fork();
			switch(child) {

				case -1:
					fprintf(stderr, APP_NAME ": failed to fork child (%s)\n", strerror(errno));
					break;

				default:

					close(tube[1]);

					monitor(tube[0], devn);

					close(tube[0]);		/* force SIGPIPE, what else can we do ? */

					for(;;) {

						dead = wait(&retn);
						if(dead == child)
							break;

						if(dead == -1 && errno != EINTR) {
							fprintf(stderr, APP_NAME ": failed waiting for child to exit (%s)\n", strerror(errno));
							return ERROR_EXIT;
						}
					}

					if(WIFSIGNALED(retn)) {
						fprintf(stderr, APP_NAME ": child exited on signal #%d\n", WTERMSIG(retn));
						return ERROR_EXIT;
					}

					return WEXITSTATUS(retn);

				case 0:

					close(tube[0]);

					/* replace -C argument with our descriptor */

					desc[0] = '0' + tube[1];
					desc[1] = '\0';
			}
		}
	}

	execvp(argp[0], argp);

	fprintf(stderr, APP_NAME ": failed to exec \"%s\" (%s)\n", argp[0], strerror(errno));

	return ERROR_EXIT;
}

/* vi:set ts=3 sw=3 cin: */
